### PR TITLE
[topgen] Sync topgen from integrated_dev to master

### PR DESCRIPTION
**Pull request by Razer6**
This PR merges the topgen tool from `integrated_dev` to master. The major part of this PR is the support for multiple address spaces by @a-will. It further includes support for generically filtering MMIO regions, and changes related to Darjeeling's device structure (more than one ROM, no flash).

Work-around changes related to the clock manager are not yet merged, since it's not yet decided how that IP will look like in master eventually.

### CODE REVIEW

File: util/topgen/templates/chiplevel.sv.tpl
Comment (vogelpi) [inline util/topgen/templates/chiplevel.sv.tpl:1178]: Is this still an issue?

File: util/topgen/validate.py
Comment (pamaury) [inline util/topgen/validate.py:205]: This setting does not seem to play a role in this PR (yet?)

**Review by vogelpi — APPROVED** [review]
Thanks for the PR @Razer6 . The hardware-related changes look good to me, the RTL isn't changed yet, just the hjson files and the tooling this is perfect.

For the software-related changes, I would like to get @pamaury's feedback as well. We should make sure these changes don't further complicate his work / if there is a way to make it go smoother.

**Comment by vogelpi** [comment]
CHANGE AUTHORIZED: hw/top_earlgrey/data/top_earlgrey.hjson
CHANGE AUTHORIZED: hw/top_earlgrey/data/xbar_main.hjson
CHANGE AUTHORIZED: hw/top_earlgrey/data/xbar_peri.hjson
CHANGE AUTHORIZED: sw/device/silicon_creator/rom/e2e/release/rom_e2e_self_hash_test.c

This PR touches some files relevant for generating RTL for Earlgrey but the RTL itself doesn't change. This is fine. The changes to the ROM self hash test consist of renaming a variable only. This is fine.

File: util/topgen/templates/chiplevel.sv.tpl
Comment (Razer6) [reply util/topgen/templates/chiplevel.sv.tpl:1178]: I think yes. In general, I don't want to make changes on these carry-over PRs.

**Comment by andreaskurth** [comment]
CHANGE AUTHORIZED: hw/top_earlgrey/data/top_earlgrey.hjson
CHANGE AUTHORIZED: hw/top_earlgrey/data/xbar_main.hjson
CHANGE AUTHORIZED: hw/top_earlgrey/data/xbar_peri.hjson
CHANGE AUTHORIZED: sw/device/silicon_creator/rom/e2e/release/rom_e2e_self_hash_test.c

Agree with @vogelpi's assessment.

**Comment by andreaskurth** [comment]
It looks like @pamaury has reviewed this and didn't request changes, though we should give @matutem a chance to chime in as well. Otherwise LGTM!

File: util/tlgen/item.py
Comment (matutem) [inline util/tlgen/item.py:39]: Should this be
`addr_ranges: Dict[int, Tuple[int, int]] = {}`

But more curious is why these classes declare both instance and class variables. It would seem the class variables are unused. Was this code just created by @a-will or are you familiar enough with it?

Comment (Razer6) [reply util/tlgen/item.py:39]: Oh yes, that's correct! 

This behavior is due to Python's typing system. The variables shown above are instance variables by default, not class variables. To explicitly declare a class variable with type hints, you need to use the `ClassVar` annotation.

I think the following documentation covers this:
https://typing.readthedocs.io/en/latest/spec/class-compat.html

File: util/topgen/validate.py
Comment (matutem) [inline util/topgen/validate.py:858]: The check in line 866 should be performed before this line, probably in line 851.

File: util/topgen/top.py
Comment (matutem) [inline util/topgen/top.py:7]: Union is used in line 71, so keep it. Also, the list of imports is sorted alphabetically, so reorder Set.

File: util/topgen/templates/xbar_tgl_excl.cfg.tpl
Comment (matutem) [inline util/topgen/templates/xbar_tgl_excl.cfg.tpl:33]: Is this actually needed? Otherwise remove the TODO.

File: util/topgen/templates/toplevel_memory.ld.tpl
Comment (matutem) [inline util/topgen/templates/toplevel_memory.ld.tpl:5]: I don't see a use of `lib` in the modified code. Is this needed?

File: util/topgen/templates/xbar_env_pkg__params.sv.tpl
Comment (matutem) [inline util/topgen/templates/xbar_env_pkg__params.sv.tpl:55]: Is this TODO really needed or just speculative? If the latter, please remove it.

File: util/topgen/lib.py
Comment (matutem) [inline util/topgen/lib.py:88]: Good replicated code removal :)

File: util/topgen.py
Comment (matutem) [inline util/topgen.py:511]: I assume this was expected to come from line 503. Is it really necessary to pop this? Seems confusing.

File: util/topgen/templates/xbar_env_pkg__params.sv.tpl
Comment (a-will) [reply util/topgen/templates/xbar_env_pkg__params.sv.tpl:55]: An issue should still be created for the missing feature (and set to an appropriate priority level, even if it's P4 "probably never"), and the limitations should be documented. It needs to be obvious that the support for multiple address spaces is _only_ for entirely disjoint networks. With the current code, two hosts using different address spaces can't connect to the same device node.

Due to this limitation, IPs need to have a separate node defined for each address space, even if the hosts would actually access the same CSRs with the same privileges (tail-wagging-dog property flow). For the current set of known use cases, this seems to be fine, however.

File: util/tlgen/lib.py
Comment (matutem) [inline util/tlgen/lib.py:159]: This is weird: if addr is a dict then int(addr, 0) should give an error since int would expect addr to be a string. The caller seems to pass a string, so what is happening with the type annotation?

File: util/tlgen/lib.py
Comment (matutem) [inline util/tlgen/lib.py:35]: What is this @param documenting? There is no addrs parameter, and Dict[ASID,Addr] seems an incorrect type.

File: util/tlgen/lib.py
Comment (matutem) [inline util/tlgen/lib.py:65]: The first arg of find_pow2_size seems to be a string?

File: util/topgen/templates/toplevel_memory.ld.tpl
Comment (Razer6) [reply util/topgen/templates/toplevel_memory.ld.tpl:5]: Not needed, the usage of `lib` was removed in this commit series. I removed the import

File: util/topgen/validate.py
Comment (Razer6) [reply util/topgen/validate.py:858]: I moved the check above and changed it to raise an exception before reaching the usage of the entry.

File: util/topgen/top.py
Comment (Razer6) [reply util/topgen/top.py]: I agree that it is used as a typing hint. Though when importing `Union`, CI fails with:
```
  flake8 util/design/gen-top-docs.py util/reggen/gen_dv.py util/tlgen/elaborate.py util/tlgen/item.py util/tlgen/lib.py util/tlgen/validate.py util/tlgen/xbar.py util/topgen.py util/topgen/c.py util/topgen/c_test.py util/topgen/gen_dv.py util/topgen/intermodule.py util/topgen/lib.py util/topgen/merge.py util/topgen/rust.py util/topgen/top.py util/topgen/validate.py
	util/topgen/top.py:7:1: F401 'typing.Union' imported but unused
```

File: util/topgen/templates/xbar_env_pkg__params.sv.tpl
Comment (Razer6) [reply util/topgen/templates/xbar_env_pkg__params.sv.tpl:55]: I created https://github.com/lowRISC/opentitan/issues/24801 to track that.

File: util/topgen/templates/xbar_tgl_excl.cfg.tpl
Comment (Razer6) [reply util/topgen/templates/xbar_tgl_excl.cfg.tpl:33]: I created https://github.com/lowRISC/opentitan/issues/24801 to track that.

File: util/tlgen/lib.py
Comment (Razer6) [reply util/tlgen/lib.py:159]: Good point. I think the type annotation was added in the meanwhile on master. I updated the annotation to let `addr` be a string. The function is only used once.

File: util/tlgen/lib.py
Comment (Razer6) [reply util/tlgen/lib.py:65]: See the comment above.

File: util/tlgen/lib.py
Comment (Razer6) [reply util/tlgen/lib.py:35]: Lets drop that comment.

File: util/topgen/top.py
Comment (matutem) [reply util/topgen/top.py]: My mistake: Union is used within a comment, so should not be imported.

File: util/topgen/top.py
Comment (matutem) [inline util/topgen/top.py:98]: Please move the type comment into the declaration, here and in line 101.

File: util/topgen/validate.py
Comment (matutem) [inline util/topgen/validate.py:854]: Just a suggestion: replace <string>.format(...) by f-stings, since they are much more readable.

File: util/topgen/rust.py
Comment (matutem) [inline util/topgen/rust.py:184]: Maybe this should be a constructor parameter?

File: util/topgen/rust.py
Comment (matutem) [inline util/topgen/rust.py:11]: Real nit: sort import items?

File: util/topgen/merge.py
Comment (matutem) [inline util/topgen/merge.py:221]: The second value in check_bool's returned tuple indicates an error, so it should be checked.

File: util/tlgen/item.py
Comment (matutem) [inline util/tlgen/item.py:71]: Nit, is this needed? The decl in line 39 sets it to {} already.

File: util/tlgen/validate.py
Comment (matutem) [inline util/tlgen/validate.py:269]: Add type, as in addr_ranges: Dict[str, [List[Tuple[int, int]]] = {}

File: util/topgen.py
Comment (matutem) [inline util/topgen.py:614]: You cannot ignore the return value from gen_dv: non-zero indicates an error.

File: util/topgen/c.py
Comment (matutem) [inline util/topgen/c.py:94]: Same as rust.py: pass addr_space as an arg?

File: util/topgen/intermodule.py
Comment (matutem) [inline util/topgen/intermodule.py:1051]: Curious: Is this a new possibility or was the code missing it?

Comment (a-will) [reply util/topgen/intermodule.py:1051]: I think you could argue either way. This is from 239f142fd09, which originally was part of https://github.com/lowRISC/opentitan-integrated/pull/280

File: util/topgen/rust.py
Comment (Razer6) [reply util/topgen/rust.py:184]: This will already be fixed in a follow-up PR: https://github.com/lowRISC/opentitan/pull/23122 For now, I would hold-off changing that in this PR.

File: util/topgen/c.py
Comment (Razer6) [reply util/topgen/c.py:94]: Same comment as above.

File: util/topgen/merge.py
Comment (Razer6) [reply util/topgen/merge.py:221]: `check_bool` already prints an error message. I also checked the error code and raise a `ValueError` if necessary.

File: util/topgen.py
Comment (Razer6) [reply util/topgen.py:614]: Changed that back to the original code.

File: util/topgen.py
Comment (Razer6) [reply util/topgen.py:642]: Since this PR is not touching that particular code, I probably want to hold off with that. I created an issue with that instead: https://github.com/lowRISC/opentitan/issues/24822

**Comment by Razer6** [comment]
@matutem Thank you for the thorough review. I changed the PR and/or added comments.

File: util/topgen.py
Comment (Razer6) [reply util/topgen.py]: 👍

File: util/topgen/merge.py
Comment (matutem) [reply util/topgen/merge.py]: Thanks

File: util/tlgen/item.py
Comment (matutem) [reply util/tlgen/item.py:71]: Cool

File: util/topgen.py
Comment (matutem) [reply util/topgen.py:641]: Okay

File: util/topgen/intermodule.py
Comment (matutem) [reply util/topgen/intermodule.py:1051]: Okay

**Review by matutem — APPROVED** [review]
Thanks for your patience :)
LGTM

**Comment by Razer6** [comment]
@andreaskurth Can we please get that merged to unblock some upstream and downstream work?

**Pull request merged**
